### DISCUSSION TITLE
gRPC Span Propagation

### DISCUSSION THREAD

**Discussion by Hablapatabla** [open-telemetry/opentelemetry-java-instrumentation#2960] [Q&A]
I am working on writing a simple instrumented client-server example using gRPC over at the opentelemetry-c++ repo. I am facing an odd issue, and I do not know if it is because of my own ignorance or not. We can get spans propagating across the network boundary in the HTTP example, as demonstrated by span output from the server containing the client root span id as the parent id. However, I simply can not get it to work with gRPC.

My worry is that propagation requires some minimal set of the HTTP2 headers associated with the gRPC call. I had quelled this worry and assumed it was my own poor implementation, but the continued difficulty led me to look at examples in other languages, and the use of gRPC interceptors in the Java example reignited that fear. In my research, I found that there is virtually no way to extract these headers in C++, as the gRPC C++ implementation still does not feature interceptors, they are an experimental feature. I am just looking for any opentelemetry-specific insight about this topic that the Java team may have.

**Comment by jkwatson** [comment]
@Hablapatabla I'm going to transfer this discussion over to the instrumentation repository, where I think it might be more appropriate.

**Reply by Hablapatabla** [reply]
Thank you!

**Comment by anuraaga** [comment]
Hi @Hablapatabla - unfortunately using the interceptors is the best bet indeed for the reasons you've found. It seems like the experimental interceptors were added precisely for OpenTracing, and they haven't seemed to prioritize making them non-experimental. @jsuereth would you be able to help champion getting this resolved so OpenTelemetry instrumentations can be marked stable unlike OpenTracing? :)

https://github.com/grpc/grpc/issues/8767

**Reply by Hablapatabla** [reply]
Well, this is certainly interesting. Thank you so much for the response! I'll have to look further into this...
